common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -489,178 +479,152 @@ public static String bytesToIpAddress(byte[] bytes, int offset, int length) {
-                        if (doubleColon) {
-                            return false;
-                        }
-                        doubleColon = true;
-                    }
-                    word.delete(0, word.length());
-                    break;
+                    j -= 5;
+                }
 
-                default:
-                    if (word != null && word.length() > 3) {
-                        return false;
-                    }
-                    if (!isValidHexChar(c)) {
+                for (; j >= start; --j) {
+                    char tmpChar = ip.charAt(j);
+                    if (tmpChar != '0' && tmpChar != ':') {
                         return false;
                     }
-                    word.append(c);
-            }
-        }
-
-        // Check if we have an IPv4 ending
-        if (numberOfPeriods > 0) {
-            // There is a test case with 7 colons and valid ipv4 this should resolve it
-            if (numberOfPeriods != 3 || !(isValidIp4Word(word.toString()) && (numberOfColons < 7 || doubleColon))) {
-                return false;
-            }
-        } else {
-            // If we're at then end and we haven't had 7 colons then there is a
-            // problem unless we encountered a doubleColon
-            if (numberOfColons != 7 && !doubleColon) {
-                return false;
-            }
+                }
 
-            if (word.length() == 0) {
-                // If we have an empty word at the end, it means we ended in either
-                // a : or a .
-                // If we did not end in :: then this is invalid
-                if (ipAddress.charAt(endOffset - 1) == ':' &&
-                    ipAddress.charAt(endOffset - 2) != ':') {
-                    return false;
+                // 7 - is minimum IPv4 address length
+                int ipv4End = ip.indexOf('%', ipv4Start + 7);
+                if (ipv4End < 0) {
+                    ipv4End = end;
                 }
-            } else if (numberOfColons == 8 && ipAddress.charAt(startOffset) != ':') {
+                return isValidIpV4Address(ip, ipv4Start, ipv4End);
+            case '%':
+                // strip the interface name/index after the percent sign
+                end = i;
+                break loop;
+            default:
                 return false;
             }
         }
 
-        return true;
+        // normal case without compression
+        if (compressBegin < 0) {
+            return colons == 7 && wordLen > 0;
+        }
+
+        return compressBegin + 2 == end ||
+               // 8 colons is valid only if compression in start or end
+               wordLen > 0 && (colons < 8 || compressBegin <= start);
     }
 
-    private static boolean isValidIp4Word(String word) {
-        char c;
-        if (word.length() < 1 || word.length() > 3) {
+    // package-visible for benchmarks
+    static boolean isValidIpV4Word(CharSequence word, int from, int toExclusive) {
```

Comment:
this is package private and used in the benchmark representing the old code. This may skew the benchmark results when comparing the new vs the old.

If these package private methods are only used from the benchmarks consider just making them private and only benchmarking the public methods.